### DISCUSSION TITLE
Bulk ideas for Monologue

### DISCUSSION THREAD

**Discussion by atomic-junky** [monologue-tool/monologue#24] [Ideas]
This discussion is about the future of Monologue, about new ideas that could make the software more interesting or more convenient to use. 
The ideas below are not necessarily feasible or even really relevant, but are just here to be discussed. The principle of this discussion is to propose lots of ideas to see which ones are interesting.

## Ideas list
- Export a project to a new Godot project (I'm really skeptical about the feasibility of this idea)
- Make MonologueProcess or Monologue an addon for Godot 4
- Support 3D models/scenes for characters
- Write MonologueProcess in different languages
- A localisation feature
- Make a better documentation/wiki
- Change the graphic identity

**Comment by RailKill** [comment]
Very exciting to see these ideas! On the subject of writing MonologueProcess in different languages/engines, I think this is much needed if we want more adoption. Users generally do not want to write their own parser/runner for the JSON, so unless they can run the dialogue in their engine like Test, they would insist that Monologue is for Godot only.


Throwing in some extra ideas, these are some features I think we can achieve in the near future:
- Character sprite/frame/animation customization
- More ActionNode features like scene transition effects, camera shakes
- Allow setting to automatically go to next once sentence/voiceline is finished


These are some crazy ones for the far future, just something to think about:
- Export dialogue as screenplay format (PDF) or excel sheet for voice actors
- Server/client for real-time collaboration to support multiple users editing at once

**Reply by atomic-junky** [reply]
I love all these ideas, I think Monologue can really become an incredible piece of software. But in my opinion, we need to completely rethink the way some of this stuff works.

I had started on the feature/graph-node-rework branch to rethink the way GraphNodes work, trying to have as few hardcoded things as possible. Maybe we should open an issue for that. But I think we need to totally rethink and split Control.gd and MonologueGraphEdit to be able to implement as many new features as possible, while avoiding as many new bugs as possible with each modification.
I think we should create a milestone for a future v3 of Monologue with the aim of making the software and code as reliable as possible.

Tell me what you think, I'd love to hear your opinion on the subject.

I also want you to know that it's a pleasure to have you working on this project. That's exactly why this project is open-source, so that people like you can come in and make changes and improvements. I think Monologue would be dead without you, so thank you so much for taking the time to make it better! ❤️

**Reply by RailKill** [reply]
Indeed there needs to be more separation of concerns especially in Control.gd, but on the brightside, I think the stuff you've implemented already have the right idea. I don't think it's that bad once we separate the logic into their own scenes and scripts so that they are more encapsulated and reusable.

I share the view that Monologue is incredible, I think it already is. I was looking for a tool that I can just pick up, make dialogues quickly and test them immediately, which was already the case before I even contributed any code to this project. So thank you as well for bringing this project to light in the first place. I don't think there is another dialogue designer tool that is as simple and intuitive to use as this one. Each node serves a very clear use case and it's elegantly powerful in my opinion.

Yeah I agree that we should focus on cleaning up the code. Localization feature is also very enticing, but I wonder how the file/data structure would look like, how we can keep Monologue's simple and elegant structure. If you need any help with specific sections, you can create an issue and I can look into it.

**Reply by atomic-junky** [reply]
Thank you! For the localisation feature I was thinking of a table containing all the sentences that can be translated, and like an Excel spreadsheet, you could translate the sentences in this table. It could take any form you like, but I was imagining a place in the options or something.
